blockstore/splitstore/splitstore.go

```diff
@@ -458,266 +616,528 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
 	return nil
 }
 
-func (s *SplitStore) warmup(curTs *types.TipSet) {
-	epoch := curTs.Height()
-
-	batchHot := make([]blocks.Block, 0, batchSize)
-	batchSnoop := make([]cid.Cid, 0, batchSize)
-
-	count := int64(0)
-	err := s.chain.WalkSnapshot(context.Background(), curTs, 1, s.skipOldMsgs, s.skipMsgReceipts,
-		func(cid cid.Cid) error {
-			count++
+// transactionally protect incoming tipsets
+func (s *SplitStore) protectTipSets(apply []*types.TipSet) {
+	s.txnLk.RLock()
+	defer s.txnLk.RUnlock()
 
-			has, err := s.hot.Has(cid)
-			if err != nil {
-				return err
-			}
+	if !s.txnActive {
+		return
+	}
 
-			if has {
-				return nil
-			}
+	var cids []cid.Cid
+	for _, ts := range apply {
+		cids = append(cids, ts.Cids()...)
+	}
 
-			blk, err := s.cold.Get(cid)
-			if err != nil {
-				return err
-			}
+	s.trackTxnRefMany(cids)
+}
 
-			batchHot = append(batchHot, blk)
-			batchSnoop = append(batchSnoop, cid)
+// transactionally protect a view
+func (s *SplitStore) protectView(c cid.Cid) {
+	s.txnLk.RLock()
+	defer s.txnLk.RUnlock()
 
-			if len(batchHot) == batchSize {
-				err = s.tracker.PutBatch(batchSnoop, epoch)
-				if err != nil {
-					return err
-				}
-				batchSnoop = batchSnoop[:0]
+	if s.txnActive {
+		s.trackTxnRef(c)
+	}
 
-				err = s.hot.PutMany(batchHot)
-				if err != nil {
-					return err
-				}
-				batchHot = batchHot[:0]
-			}
+	s.txnViewsMx.Lock()
+	s.txnViews++
+	s.txnViewsMx.Unlock()
+}
 
-			return nil
-		})
+func (s *SplitStore) viewDone() {
+	s.txnViewsMx.Lock()
+	defer s.txnViewsMx.Unlock()
 
-	if err != nil {
-		log.Errorf("error warming up splitstore: %s", err)
-		return
+	s.txnViews--
+	if s.txnViews == 0 && s.txnViewsWaiting {
+		s.txnViewsCond.Signal()
 	}
+}
 
-	if len(batchHot) > 0 {
-		err = s.tracker.PutBatch(batchSnoop, epoch)
-		if err != nil {
-			log.Errorf("error warming up splitstore: %s", err)
-			return
-		}
-
-		err = s.hot.PutMany(batchHot)
-		if err != nil {
-			log.Errorf("error warming up splitstore: %s", err)
-			return
-		}
-	}
+func (s *SplitStore) viewWait() {
+	s.txnViewsMx.Lock()
+	defer s.txnViewsMx.Unlock()
 
-	if count > s.markSetSize {
-		s.markSetSize = count + count>>2 // overestimate a bit
+	s.txnViewsWaiting = true
+	for s.txnViews > 0 {
+		s.txnViewsCond.Wait()
 	}
+	s.txnViewsWaiting = false
```

Comment:
So, we're waiting for all the views to end, but then we're not _doing_ anything atomically. That can't be right.

Should we be waiting for all the non-tracked views to end? I.e., should we have a return at https://github.com/filecoin-project/lotus/pull/6474/files#diff-eac9e730a0594047de6e81aa421dcd33aac194e505a18945ca45e72db789687cR642?

Reply:
wait, what?  we have already started the transaction.
All the views are tracked now as they always increment the rlock.
The bariier just ensures that there is no view that started before the transaction that hasn't ended.

Reply:
Ok. This _should_ be fine.

Ideally, we'd track views started after the transaction starts separately, but that's not strictly necessary.

Reply:
Yeah.

To summarize our sync conversation: a long running view is a bug and i'd very much rather hang the compaction (which is something we'll see in the logs) than do a potentially catastrophic delete.